src/index.ts

```diff
@@ -591,22 +471,86 @@ class LayercodeClient implements ILayercodeClient {
     return this.wavRecorder.getStream();
   }
 
+  async setInputDevice(deviceId: string): Promise<void> {
+    try {
+      this.deviceId = deviceId;
+
+      // Restart recording with the new device
+      await this._restartAudioRecording();
+
+      // Reinitialize VAD with the new audio stream if VAD is enabled
+      if (this.vadManager.isVADEnabled()) {
+        console.log('Reinitializing VAD with new audio stream');
+        const newStream = this.wavRecorder.getStream();
+        this.vadManager.reinitialize(newStream);
+      }
+
+      // Set up amplitude monitoring
+      this._setupAmplitudeMonitoring(this.wavRecorder, this.options.onUserAmplitudeChange, (amp) => (this.userAudioAmplitude = amp));
+
+      console.log(`Successfully switched to input device: ${deviceId}`);
+    } catch (error) {
+      console.error(`Failed to switch to input device ${deviceId}:`, error);
+      throw new Error(`Failed to switch to input device: ${error instanceof Error ? error.message : String(error)}`);
+    }
+  }
+
   /**
-   * Switches the input device for the microphone and restarts recording
-   * @param {string} deviceId - The deviceId of the new microphone
+   * Manually switches to the next available audio input device
    */
-  async setInputDevice(deviceId: string): Promise<void> {
-    if (this.wavRecorder) {
-      try {
-        await this.wavRecorder.end();
-      } catch (e) {}
+  private async _switchToNextDevice(): Promise<void> {
+    try {
+      const devices = await this.wavRecorder.listDevices();
+      const audioInputs = devices.filter((d) => d.kind === 'audioinput');
+      const nextDevice = audioInputs.find((d) => d.deviceId !== this.deviceId);
+
+      if (nextDevice) {
+        await this.setInputDevice(nextDevice.deviceId);
+      }
+    } catch (error) {
+      console.error('Error switching to next device:', error);
+      throw error;
+    }
+  }
+
+  /**
+   * Sets up the device change event listener
+   */
+  private _setupDeviceChangeListener(): void {
+    this.wavRecorder.listenForDeviceChange(async (devices) => {
       try {
-        await this.wavRecorder.quit();
-      } catch (e) {}
+        // Ensure we have current device ID
+        if (!this.deviceId) {
+          this._refreshCurrentDeviceId();
+        }
+
+        const currentDeviceExists = devices.some((device) => device.deviceId === this.deviceId);
+        if (!currentDeviceExists && this.deviceId) {
+          await this._switchToNextDevice();
```

Comment:
Since _switchToNextDevice is only ever called, here, we could just put the few lines of code it has right here.